### PR TITLE
[C++] Content formatting

### PR DESCRIPTION
**Pull request by rachel-mack**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-java/blob/master/REVIEWING.md)

This is meant to be a direct copy of the existing C++ docs, reformatted into .rst, in a TBXD repo.
[Original source](https://github.com/mongodb/mongo-cxx-driver)
[Original site](https://mongocxx.org/)

JIRA - <https://jira.mongodb.org/browse/DOCSP-35153>
Staging - <https://docs-mongodbcom-staging.corp.mongodb.com/cpp-driver/docsworker-xlarge/content-formatting/>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/installation.txt [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+Advanced Options
+----------------
+
+- :ref:`Advanced Configuration and Installation Options <cpp-installation-advanced>`
+
+Package Managers
+----------------
+
+The Mongo C++ driver is available in the following package managers.
+- `Vcpkg <https://vcpkg.io/>`__ (search for mongo-cxx-driver)
```
Comment (norareidy) [inline source/installation.txt:59]: I: add a newline before the first bullet point so the list renders correctly
```suggestion
The Mongo C++ driver is available in the following package managers:

- `Vcpkg <https://vcpkg.io/>`__ (search for mongo-cxx-driver)
```

File: source/installation.txt [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+If you do not already have Conan installed, then install it and run the Conan
+initalization command below:
+
+.. code-block:: bash
+
+   $ pip install conan
+   $ conan profile detect --force
+
+Add the following to your ``conanfile.txt``
```
Comment (norareidy) [inline source/installation.txt:119]: S: add a colon
```suggestion
Add the following to your ``conanfile.txt``:
```

File: source/installation/advanced.txt [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+       -DCMAKE_CXX_STANDARD=11 \
+       -Bcmake-build -S.
+
+   cmake --build cmake-build --target install
+   $HOME/app/bin/app.out
+   # Prints "successfully connected with C++ driver"
+
+See the cmake documentation for [RPATH handling](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling) for more information.
+
+### Fixing the "cannot open shared object file" error on Linux
```
Comment (norareidy) [inline source/installation/advanced.txt:258]: I: header formatting
```suggestion
Fixing the "cannot open shared object file" error on Linux
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: source/installation/advanced.txt [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+       -DCMAKE_INSTALL_PREFIX=$HOME/app \
+       -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=TRUE \
+       -DCMAKE_CXX_STANDARD=11 \
+       -Bcmake-build -S.
+
+   cmake --build cmake-build --target install
+   $HOME/app/bin/app.out
+   # Prints "successfully connected with C++ driver"
+
+See the cmake documentation for [RPATH handling](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling) for more information.
```
Comment (norareidy) [inline source/installation/advanced.txt:256]: I: link formatting
```suggestion
See the cmake documentation for `RPATH handling <https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling>`_ for more information.
```

File: source/polyfill-selection.txt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+.. toctree::
+   :titlesonly:
+   :maxdepth: 1
+
+The mongocxx driver uses C++17 features ``std::optional`` and ``std::string_view``.
+To use the C++17 standard library implementations for these features, set
+the CMake configuration variable ``CMAKE_CXX_STANDARD`` to 17 or higher.
+Otherwise, to compile the mongocxx driver for pre-C++17 configurations, a
+polyfill library implementation must be selected from the following options
```
Comment (norareidy) [inline source/polyfill-selection.txt:25]: I: add colon
```suggestion
polyfill library implementation must be selected from the following options:
```

File: source/polyfill-selection.txt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  will be set to ON by default in an upcoming major release).
+
+- MNMLSTC/core (*default for non-Windows platforms*)
+
+  **This option is deprecated and will be removed in an upcoming major release.**
+  Select with ``-DBSONCXX_POLY_USE_MNMLSTC=1``. **NOTE**: This option vendors a
+  header-only installation of MNMLSTC/core into the bsoncxx library installation
+  and will therefore download MLNMLSTC from GitHub during the configuration
+  process. If you already have an available version of MNMLSTC on your system,
+  you can avoid the download step by using ``-DBSONCXX_POLY_USE_SYSTEM_MNMLSTC``.
```
Comment (norareidy) [inline source/polyfill-selection.txt:42]: S: I think using the note directive here makes sense
```suggestion
  Select with ``-DBSONCXX_POLY_USE_MNMLSTC=1``. 
  .. note::
  
     This option vendors a header-only installation of MNMLSTC/core into the bsoncxx 
     library installation and will therefore download MLNMLSTC from GitHub during the 
     configuration process. If you already have an available version of MNMLSTC on your 
     system, you can avoid the download step by using ``- 
     DBSONCXX_POLY_USE_SYSTEM_MNMLSTC``.
```

File: source/polyfill-selection.txt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+- Boost (*default for Windows platforms*)
+
+  **This option is deprecated and will be removed in an upcoming major release.**
+  Select with ``-DBSONCXX_POLY_USE_BOOST=1``. This is currently the only
+  non-bsoncxx option if you are using a version of MSVC that does not support
+  C++17.
+
+.. note::
+
+   "default" refers to **pre-C++17** configurations when no polyfill library is explicitly selected):
```
Comment (norareidy) [inline source/polyfill-selection.txt:53]: I: remove extra colon
```suggestion
   "default" refers to **pre-C++17** configurations when no polyfill library is explicitly selected)
```

File: source/testing.txt [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+   ctest -R bson
+
+could be used to run only the bson tests.
+
+Running Integration Tests
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some of the tests require a running mongod instance.  For this, first download
+the [mongodb server](https://www.mongodb.com/download-center).
```
Comment (norareidy) [inline source/testing.txt:69]: I: link formatting
```suggestion
the `mongodb server <https://www.mongodb.com/download-center>`_.
```

File: source/tutorial.txt [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+----------------
+
+To update documents in a collection, you can use the collection's
+``update_one()`` and ``update_many()`` methods.
+
+The update methods return an instance of ``std::optional<
+mongocxx::result::update >``, which provides information about the operation
+including the number of documents modified by the update. For more information,
+see `mongocxx::result::update
+<{+api+}/classmongocxx_1_1v__noabi_1_1result_1_1update>`__
```
Comment (norareidy) [inline source/tutorial.txt:396]: I: add period
```suggestion
<{+api+}/classmongocxx_1_1v__noabi_1_1result_1_1update>`__.
```

File: source/tutorial.txt [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+Delete Documents
+----------------
+
+To delete documents from a collection, you can use a collection's
+``delete_one()`` and ``delete_many()`` methods.
+
+The delete methods return an instance of ``std::optional<
+mongocxx::result::delete >``, which contains the number of documents deleted.
+For more information, see `mongocxx::result::delete
+<{+api+}/classmongocxx_1_1v__noabi_1_1result_1_1delete__result>`__
```
Comment (norareidy) [inline source/tutorial.txt:440]: i: period
```suggestion
<{+api+}/classmongocxx_1_1v__noabi_1_1result_1_1delete__result>`__.
```

File: source/working-with-bson.txt [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+   if (store_ele) {
+       // this block will only execute if "store" was found in the document
+       std::cout << "Examining inventory at " << to_json(store_ele.get_value()) << std::endl;
+   }
+
+This feature is shown in more detail in `this example
+<https://github.com/mongodb/mongo-cxx-driver/blob/master/examples/bsoncxx/getting_values.cpp>`__
+and `this example
+<https://github.com/mongodb/mongo-cxx-driver/blob/master/examples/bsoncxx/view_and_value.cpp>`__.
+
```
Comment (norareidy) [inline source/working-with-bson.txt:376]: S: add an anchor so you can include this section in the overview

```suggestion
.. _cpp-bson-types:

```

File: source/working-with-bson.txt [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+`examples <https://github.com/mongodb/mongo-cxx-driver/tree/master/examples/bsoncxx>`__.
+
+1. :ref:`Document Builders <cpp-bson-builders>`
+#. :ref:`Owning BSON Documents (values) <cpp-bson-value>`
+#. :ref:`Non-owning BSON Documents (views) <cpp-bson-view>`
+#. :ref:`Optionally-owning BSON Documents(view_or_value) <cpp-bson-view_or_value>`
+#. :ref:`BSON Document Lifetime <cpp-bson-lifetime>`
+#. :ref:`Printing BSON Documents <cpp-bson-print>`
+#. :ref:`Getting Fields out of BSON Documents <cpp-bson-fields>`
+
```
Comment (norareidy) [inline source/working-with-bson.txt:32]: I: the last section (BSON types) is missing from this list
```suggestion
#. :ref:`BSON Types <cpp-bson-types>`
```

File: source/working-with-bson.txt [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+   statement within the loop body should assign its result back to the in_array
+   object, so that the loop restarts in a consistent state:
+
+.. code-block:: c++
+
+   for (auto && e : {1, 2, 3}) {
+       auto open_state = in_array << builder::stream::open_document;
+       auto temp_state = open_state << "key" << e;
+       in_array = temp_state << builder::stream::close_document;
+   }
```
Comment (norareidy) [inline source/working-with-bson.txt:195]: S: since this code block is described in the note above, I think it should be moved inside the note directive

**Review by norareidy — CHANGES_REQUESTED** [review]
Added some suggestions, mostly rST format-related!

File: source/contributing.txt [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+General structure:
+
+- License
+- Include Guard (``#pragma once``)
+- Header Prelude
+- System Headers ``<vector>`` (alphabetical order)
+- Driver Headers ``<path/to/header.hpp>`` (alphabetical order)
+- Open Namespace mongocxx
+- ``inline namespace v_noabi {``
+-    Code
```
[the remark this answers is not in the document]
Comment (rachel-mack) [reply source/contributing.txt:77]: This is for formatting in the printed text - it's meant to look like the `Code` is nested between the `{` and `}` on the previous and following lines.

File: source/working-with-bson.txt [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+      }
+
+.. _cpp-bson-value:
+
+Owning BSON Documents
+~~~~~~~~~~~~~~~~~~~~~
+
+`bsoncxx::document::value <https://github.com/mongodb/mongo-cxx-driver/blob/master/src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/document/value.hpp>`__
+
+This type represents an actual BSON document, one that owns its buffer of
```
Comment (norareidy) [inline source/working-with-bson.txt:205]: S: not a necessary change, but this link seems a little out of place - might be helpful to link some of the body text instead
```suggestion

The `bsoncxx::document::value type <https://github.com/mongodb/mongo-cxx-driver/blob/master/src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/document/value.hpp>`__ represents an actual BSON document, one that owns its buffer of
```

Comment (norareidy) [reply source/working-with-bson.txt:205]: Applies to the Non-owning BSON Documents (views) and Printing BSON Documents sections as well

File: source/working-with-bson.txt [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+
+   document::value ping = document{} << "ping" << 1 << finalize;
+
+   // You can pass a document::view into run_command()
+   db.run_command(ping.view());
+
+   // Or you can move in a document::value
+   db.run_command(std::move(ping));
+
+You shouldn't need to create view_or_value types directly in order to use
```
Comment (norareidy) [inline source/working-with-bson.txt:287]: S: change to monospace format
```suggestion
You shouldn't need to create ``view_or_value`` types directly in order to use
```

File: source/working-with-bson.txt [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+
+   auto collection2 = db["cookbooks"];
+   auto cursor2 = collection2.find(query_value.view());
+
+.. _cpp-bson-view_or_value:
+
+Optionally-owning BSON Documents (view_or_value)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Many driver methods take a document::view_or_value parameter, for example,
```
Comment (norareidy) [inline source/working-with-bson.txt:266]: S: small change (that's maybe outside the scope of this review), but I think all mentions of the `document::` types should be monospaced
```suggestion
Many driver methods take a ``document::view_or_value`` parameter, for example,
```

File: source/testing.txt [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+   ctest -R bson
+
+could be used to run only the bson tests.
+
+Running Integration Tests
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some of the tests require a running mongod instance.  For this, first download
+the `mongodb server <https://www.mongodb.com/download-center>`__.
```
Comment (norareidy) [inline source/testing.txt:69]: S: capitalization
```suggestion
the `MongoDB server <https://www.mongodb.com/download-center>`__.
```

File: source/contributing.txt [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+Class Declarations
+~~~~~~~~~~~~~~~~~~
+
+Guidelines:
+
+- Blank line at beginning and end of class declaration
+- Public section up top / private at bottom
+- Lifecycle methods first (see rules above)
+- Private Member Ordering
+  - Friendships
```
Comment (norareidy) [inline source/contributing.txt:133]: I: add a newline here so the list renders correctly
```suggestion
- Private Member Ordering

  - Friendships
```

File: source/getting-help.txt [hunk 20]
```diff
@@ -0,0 +1,28 @@
+.. _cpp-getting-help:
```
Comment (norareidy) [inline source/getting-help.txt:1]: I: I don't think this page is accessible from the TOC

File: source/client-side-encryption.txt [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+Examples
+--------
+
+Automatic Client-Side Field Level Encryption
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Automatic Client-Side Field Level Encryption is enabled by creating a ``mongocxx::client`` with the ``auto_encryption_opts`` option set to an instance of ``mongocxx::options::auto_encryption``. The following examples show how to set up automatic client-side field level encryption using the ``mongocxx::client_encryption`` class to create a new encryption data key.
+
+.. note::
+   Automatic client-side field level encryption requires MongoDB 4.2 enterprise or a MongoDB 4.2 Atlas cluster. The community version of the server supports automatic decryption as well as explicit client-side field level encryption.
```
Comment (norareidy) [inline source/client-side-encryption.txt:68]: I: add a newline here
```suggestion
.. note::

   Automatic client-side field level encryption requires MongoDB 4.2 enterprise or a MongoDB 4.2 Atlas cluster. The community version of the server supports automatic decryption as well as explicit client-side field level encryption.
```

File: source/tutorial.txt (hunk 9 above)
Comment (norareidy) [reply source/tutorial.txt:396]: Bumping this comment (also applies to line 440)

**Review by norareidy — APPROVED** [review]
LGTM with some final comments, let me know if you want me to take another look!

File: source/working-with-bson.txt (hunk 15 above)
Comment (rachel-mack) [reply source/working-with-bson.txt:205]: I'm going to leave it in for now, and when we're standardizing, we can decide if it should be removed, or moved somewhere else instead.

File: source/installation.txt [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+   /installation/macos
+   /installation/linux
+   /installation/advanced
+
+Prerequisites
+-------------
+
+- Any standard Unix platform, or Windows 7 SP1+
+- A compiler that supports C++11 (gcc, clang, or Visual Studio)
+- `CMake <https://cmake.org>`__ 3.2 or later
```
Comment (kevinAlbs) [inline source/installation.txt:34]: Replace "3.2" with "3.15".

This is notably out-dated in the current C++ driver documentation. The minimum required CMake was increased to 3.15. https://github.com/mongodb/mongo-cxx-driver/pull/1116 has been opened to update the C++ driver documentation.

File: source/installation/windows.txt [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+
+   'C:\Program Files (x86)\CMake\bin\cmake.exe' .. \
+       -G "Visual Studio 15 2017" -A "x64"         \
+       -DCMAKE_CXX_STANDARD=17                     \
+       -DCMAKE_INSTALL_PREFIX=C:\mongo-cxx-driver  \
+
+Configuring with ``mongocxx`` 3.7.0 and Older
+`````````````````````````````````````````````
+
+To build versions 3.7.0 and older without a C++17 polyfill, it is necessary to configure with additional options:
```
Comment (kevinAlbs) [inline source/installation/windows.txt:70]: Add newline? Rendering does not include bullets [on staging](https://docs-mongodbcom-staging.corp.mongodb.com/cpp-driver/docsworker-xlarge/content-formatting/installation/windows/):

Screen Shot 2024-03-12 at 3 17 01 PM

**Review by kevinAlbs — COMMENTED** [review]
LGTM with formatting changes and one text replacement.

File: source/api-abi-versioning.txt [hunk 24]
```diff
@@ -0,0 +1,142 @@
+.. _cpp-api-abi-versioning:
+
+======================
+API and ABI Versioning
```
[the remark this answers is not in the document]
Comment (rachel-mack) [reply source/api-abi-versioning.txt:4]: Yup, that's fine. I can prioritize that ticket as well, if you want it in quickly.

[the remark this answers is not in the document]
Comment (kevinAlbs) [reply source/api-abi-versioning.txt:4]: Filed [DOCSP-37650](https://jira.mongodb.org/browse/DOCSP-37650)

> I can prioritize that ticket as well, if you want it in quickly.

No need. I would only request it be applied before the documentation on mongocxx.org is retired for mongodb.com documentation.

**Pull request merged**